diff --git a/gpxpy/geo.py b/gpxpy/geo.py
index 00fa3bb..a538f76 100644
--- a/gpxpy/geo.py
+++ b/gpxpy/geo.py
@@ -161,10 +161,10 @@ def calculate_max_speed(speeds_and_distances: List[Tuple[float, float]], extreem
 
     distances = [x[1] for x in speeds_and_distances]
     average_distance = sum(distances) / size
-    standard_distance_deviation = mod_math.sqrt(sum((distance - average_distance) ** 2 for distance in distances) / size)
+    standard_distance_deviation = mod_math.sqrt(sum((distance - average_distance) ** 2 for distance in distances) / (size - 1))
 
     # Ignore items where the distance is too big:
-    filtered_speeds_and_distances = [x for x in speeds_and_distances if abs(x[1] - average_distance) <= standard_distance_deviation * 1.5]
+    filtered_speeds_and_distances = [x for x in speeds_and_distances if abs(x[1] - average_distance) < standard_distance_deviation * 1.5]
 
     # sort by speed:
     speeds = [x[0] for x in filtered_speeds_and_distances]
@@ -175,7 +175,7 @@ def calculate_max_speed(speeds_and_distances: List[Tuple[float, float]], extreem
     # Even here there may be some extremes => ignore the last 5%:
     index = int(len(speeds) * (1-extreemes_percentile))
     if index >= len(speeds):
-        index = -1
+        index = len(speeds) - 1
 
     return speeds[index]
 
