diff --git a/gpxpy/geo.py b/gpxpy/geo.py
index 00fa3bb..39349fa 100644
--- a/gpxpy/geo.py
+++ b/gpxpy/geo.py
@@ -204,25 +204,22 @@ def calculate_uphill_downhill(elevations: List[Optional[float]]) -> Tuple[float,
 
     def __filter(n: int) -> float:
         current_ele = elevations[n]
-        if current_ele is None:
-            return False
         if 0 < n < size - 1:
             previous_ele = elevations[n-1]
             next_ele = elevations[n+1]
             if previous_ele is not None and current_ele is not None and next_ele is not None:
-                return previous_ele*.3 + current_ele*.4 + next_ele*.3
+                return previous_ele*.4 + current_ele*.2 + next_ele*.4
         return current_ele
 
     smoothed_elevations = list(map(__filter, range(size)))
 
     uphill, downhill = 0., 0.
     for prev, cur in zip(smoothed_elevations, smoothed_elevations[1:]):
-        if prev is not None and cur is not None:
-            d = cur - prev
-            if d > 0:
-                uphill += d
-            else:
-                downhill -= d
+        d = prev - cur
+        if d > 0:
+            uphill += d
+        else:
+            downhill -= d
     return uphill, downhill
 
 
